mlflow/utils/environment.py

```diff
@@ -729,6 +732,108 @@ def _get_pip_install_mlflow():
         return f"pip install mlflow=={VERSION} 1>&2"
 
 
+def _add_requirements_to_file(
+    new_reqs: List[Requirement],
+    file_path: str,
+) -> None:
+    """
+    Adds new requirements to the given requirements file (either `conda.yaml`
+    or `requirements.txt`). If any of the new requirements overlap with the
+    requirements in the file, the new requirements will overwrite the old ones.
+
+    Args:
+        new_reqs (List[`Requirement`]): List of new requirements to add to the file.
+        file_path (str): Path to the locally-saved requirements file. Must point to
+            either a `conda.yaml` or `requirements.txt` file.
+    """
+    old_reqs = _get_requirements_from_file(file_path)
+    updated_reqs = _add_or_overwrite_requirements(new_reqs, old_reqs)
+    _write_requirements_to_file(file_path, updated_reqs)
+    _logger.info(
+        "Done adding requirements!\n\n"
+        f"Old requirements:\n{pformat([str(req) for req in old_reqs])}\n\n"
+        f"Updated requirements:\n{pformat(updated_reqs)}\n"
+    )
+
+
+def _remove_requirements_from_file(
+    reqs_to_remove: List[Requirement],
+    file_path: str,
+) -> None:
+    """
+    Removes requirements from the given requirements file (either `conda.yaml`
+    or `requirements.txt`). If any of the provided requirements are not present
+    in the file, they will be ignored. This function also ignores version specifiers.
+    For example, if the file contains "pandas==1.2.3", and the provided requirement
+    is "pandas", the function will remove "pandas==1.2.3" from the file.
+
+    Args:
+        reqs_to_remove (List[`Requirement`]): List of requirements to remove from the file
+        file_path (str): Path to the locally-saved requirements file. Must point to
+            either a `conda.yaml` or `requirements.txt` file.
+    """
+    old_reqs = _get_requirements_from_file(file_path)
+    updated_reqs = _remove_requirements(reqs_to_remove, old_reqs)
+    _write_requirements_to_file(file_path, updated_reqs)
+    _logger.info(
+        "Done removing requirements!\n\n"
+        f"Old requirements:\n{pformat([str(req) for req in old_reqs])}\n\n"
+        f"Updated requirements:\n{pformat(updated_reqs)}\n"
+    )
+
+
+def _get_requirements_from_file(
+    file_path: str,
+) -> List[Requirement]:
+    path = pathlib.Path(file_path)
+    data = path.read_text()
+    if path.name == _CONDA_ENV_FILE_NAME:
+        conda_env = yaml.safe_load(data)
+        reqs = _get_pip_deps(conda_env)
+    else:
+        reqs = data.splitlines()
```

Comment:
Any chance that `reqs` contains comments?

Reply:
ooo good question. i'm not sure—i don't think we add them when we save models, but there's always the chance that the user manually edited it to add comments. i can maybe just ignore anything in requirements.txt or conda.yaml that doesn't successfully parse as a requirement. the downside here is that the comments will not be retained when i write back to the file.

however i wonder if we should care about this case... if the user manually edited their file, surely they can manually edit it again to remove the comments

Reply:
This is a minor edge case. Most users don't edit requirements.txt or conda.yaml manually. I think that's to support this later if there is a request.

Reply:
```python

for line in old_reqs_contents.splitlines():
    if not line.strip() or line.startswith("#"):
        yield line
    
    try:
        req = Requirement(line)
    except ...:
        yield line

    if operation == "delete" and req.name in to_delete:
        continue

    if operation == "add" and new := to_add.get(req.name):
        yield str(new)
```

Reply:
will leave this as it is for now since it may complicate the implementation for the `conda.yaml` file (we would have to do some line-by-line comparison of the output of `yaml.dump` vs. the original file), but i'll keep this reference in case anyone requests the feature in the future!

Reply:
Sounds good!